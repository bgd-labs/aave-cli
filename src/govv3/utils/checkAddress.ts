import { findObjectPaths } from 'find-object-paths';
import { Address, getAddress } from 'viem';
import * as addresses from '@bgd-labs/aave-address-book';

/**
 * Checks if address is listed on address-book
 * @param value
 * @param chainId
 * @returns string[] found paths to address-book addresses
 */
export function isKnownAddress(value: Address, chainId: number): string[] | void {
  // glob imports have no object properties
  // therefore we recreate the object via spread & remove addresses unrelated to the chain we are checking
  const transformedAddresses = Object.keys(addresses).reduce((acc, key) => {
    if ((addresses as any)[key].CHAIN_ID == chainId) acc[key] = { ...(addresses as any)[key] };
    return acc;
  }, {} as { [key: string]: any });
  // while on address book we have checksummed addresses tenderly returns non check summed
  // therefore we checksum the needle to have exact matches
  const results = findObjectPaths(transformedAddresses, { value: getAddress(value) });
  if (typeof results === 'string') return [results];
  return results;
}

export function findPayloadsController(chainId: number): Address | void {
  const key = Object.keys(addresses).find(
    (key) => addresses[key].CHAIN_ID === chainId && addresses[key].PAYLOADS_CONTROLLER
  );
  if (key) return addresses[key].PAYLOADS_CONTROLLER;
}
