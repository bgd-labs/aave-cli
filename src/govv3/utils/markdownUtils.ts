import { Hex, PublicClient } from 'viem';
import { CheckResult } from '../checks/types';

export function boolToMarkdown(value: boolean) {
  if (value) return `:white_check_mark:`;
  return `:sos:`;
}

/**
 * Turns a plaintext address into a link to etherscan page of that address
 * @param address to be linked
 * @param code whether to link to the code tab
 */
export function toAddressLink(address: Hex, md: boolean, client?: PublicClient): string {
  if (!client) return address;
  const link = `${client.chain?.blockExplorers?.default.url}/address/${address}`;
  if (md) return toMarkdownLink(link, address);
  return link;
}

/**
 * Turns a plaintext address into a link to etherscan page of that address
 * @param address to be linked
 * @param code whether to link to the code tab
 */
export function toTxLink(txn: Hex, md: boolean, client?: PublicClient): string {
  if (!client) return txn;
  const link = `${client.chain?.blockExplorers?.default.url}/tx/${txn}`;
  if (md) return toMarkdownLink(link, txn);
  return link;
}

export function toMarkdownLink(link: string, title?: any) {
  return `[${title || link}](${link})`;
}

export function renderCheckResult(check: { name: string }, result: CheckResult) {
  let response = `### Check: ${check.name} ${boolToMarkdown(!result.errors.length)}\n\n`;
  if (result.errors.length) response += `#### Errors\n\n${result.errors.join('\n')}\n\n`;
  if (result.warnings.length) response += `#### Warnings\n\n${result.warnings.join('\n')}\n\n`;
  if (result.info.length) response += `#### Info\n\n${result.info.join('\n')}\n\n`;
  return response;
}

export function renderUnixTime(time: number) {
  return new Date(time * 1000).toLocaleString('en-GB', { timeZone: 'UTC' });
}
