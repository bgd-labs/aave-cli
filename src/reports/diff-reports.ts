import hash from 'object-hash';
import { fetchRateStrategyImage } from './fetch-IR-strategy';
import { renderReserve, renderReserveDiff } from './reserve';
import { renderStrategy, renderStrategyDiff } from './strategy';
import { AaveV3Snapshot, AaveV3Reserve } from './snapshot-types';
import { diff } from './diff';
import { renderEmode, renderEmodeDiff } from './emode';

export async function diffReports<A extends AaveV3Snapshot, B extends AaveV3Snapshot>(pre: A, post: B) {
  const chainId = pre.chainId;
  const diffResult = diff(pre, post);
  // download interest rate strategies
  // only downloads if it doesn't exist yet
  for (const ir in pre.strategies) {
    await fetchRateStrategyImage(pre.strategies[ir]);
  }
  for (const ir in post.strategies) {
    await fetchRateStrategyImage(post.strategies[ir]);
  }

  // create report
  let content = '';
  const reservesAdded = Object.keys(diffResult.reserves)
    .map((reserveKey) => {
      // to being present on reserve level % trueish means reserve was added
      if ((diffResult.reserves[reserveKey] as any).to) {
        let report = renderReserve((diffResult.reserves[reserveKey] as any).to, chainId);
        const imageHash = hash(
          post.strategies[((diffResult.reserves[reserveKey] as any).to as AaveV3Reserve).interestRateStrategy]
        );
        report += renderStrategy(
          post.strategies[((diffResult.reserves[reserveKey] as any).to as AaveV3Reserve).interestRateStrategy]
        );

        report += `| interestRate | ![ir](/.assets/${imageHash}.svg) |\n`;
        if (post.reserves[reserveKey].eModeCategory && post.reserves[reserveKey].eModeCategory != 0) {
          report += renderEmode(post.eModes[post.reserves[reserveKey].eModeCategory]);
        }

        return report;
      }
    })
    .filter((i) => i);
  const reservesRemoved = Object.keys(diffResult.reserves)
    .map((reserveKey) => {
      // from being present on reserve level % trueish means reserve was removed
      if ((diffResult.reserves[reserveKey] as any).from) {
        return renderReserve((diffResult.reserves[reserveKey] as any).from, chainId);
      }
    })
    .filter((i) => i);
  const reservesAltered = Object.keys(diffResult.reserves)
    .map((reserveKey) => {
      // from being present on key means reserve was removed
      if (
        !(diffResult.reserves[reserveKey] as any).hasOwnProperty('from') &&
        Object.keys(diffResult.reserves[reserveKey]).find(
          (fieldKey) => typeof (diffResult.reserves as any)[reserveKey][fieldKey] === 'object'
        )
      ) {
        let report = renderReserveDiff(diffResult.reserves[reserveKey] as any, chainId);
        if (diffResult.reserves[reserveKey].interestRateStrategy.hasOwnProperty('from')) {
          report += renderStrategyDiff(
            diff(
              pre.strategies[(diffResult.reserves[reserveKey].interestRateStrategy as any).from],
              post.strategies[(diffResult.reserves[reserveKey].interestRateStrategy as any).to]
            ) as any
          );
          report += `| interestRate | ![before](/.assets/${hash(
            pre.strategies[(diffResult.reserves[reserveKey].interestRateStrategy as any).from]
          )}.svg) | ![after](/.assets/${hash(
            post.strategies[(diffResult.reserves[reserveKey].interestRateStrategy as any).to]
          )}.svg) |`;
        }
        if (diffResult.reserves[reserveKey].eModeCategory?.hasOwnProperty('from')) {
          report += renderEmodeDiff(
            diff(
              pre.eModes[(diffResult.reserves[reserveKey].eModeCategory as any).from] || {},
              post.eModes[(diffResult.reserves[reserveKey].eModeCategory as any).to]
            ) as any
          );
        }

        return report;
      }
    })
    .filter((i) => i);
  if (reservesAdded.length || reservesRemoved.length || reservesAltered.length) {
    content += '## Reserve changes\n\n';
    if (reservesAdded.length) {
      content += `### ${reservesAdded.length > 1 ? 'Reserve' : 'Reserves'} added\n\n`;
      content += reservesAdded.join('\n\n');
      content += '\n\n';
    }

    if (reservesAltered.length) {
      content += `### ${reservesAltered.length > 1 ? 'Reserve' : 'Reserves'} altered\n\n`;
      content += reservesAltered.join('\n\n');
      content += '\n\n';
    }

    if (reservesRemoved.length) {
      content += `### ${reservesRemoved.length > 1 ? 'Reserve' : 'Reserves'} removed\n\n`;
      content += reservesRemoved.join('\n\n');
      content += '\n\n';
    }
  }

  content += `## Raw diff\n\n\`\`\`json\n${JSON.stringify(diff(pre, post, true), null, 2)}\n\`\`\``;
  return content;
}
