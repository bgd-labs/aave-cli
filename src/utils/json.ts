import JSONbig from "json-bigint";
import fs from "fs";
import path from "path";

export function readJson(filePath: string) {
  const content = fs.readFileSync(path.join(process.cwd(), filePath));
  return JSON.parse(
    JSON.stringify(JSONbig({ storeAsString: true }).parse(content))
  );
}
