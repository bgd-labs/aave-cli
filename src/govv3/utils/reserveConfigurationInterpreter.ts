import * as pools from '@bgd-labs/aave-address-book';
import {getBits} from '../../utils/storageSlots';

export function getDecodedReserveData(contractAddress: string, data?: bigint) {
  if (!data) return data;
  if (
    [
      pools.AaveV2EthereumAMM.POOL,
      pools.AaveV2Ethereum.POOL,
      pools.AaveV2Polygon.POOL,
      pools.AaveV2Avalanche.POOL,
    ]
      .map((address) => address.toLowerCase())
      .includes(contractAddress.toLowerCase())
  )
    return decodeReserveDataV2(data);
  return decodeReserveDataV3(data);
}

function decodeReserveDataV2(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const stableBorrowingEnabled = Number(getBits(data, 59n, 59n));
  const reserveFactor = getBits(data, 64n, 79n);
  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    stableBorrowingEnabled: !!stableBorrowingEnabled,
    reserveFactor,
  };
}

function decodeReserveDataV3(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const stableRateBorrowingEnabled = Number(getBits(data, 59n, 59n));
  const paused = Number(getBits(data, 60n, 60n));
  const borrowingInIsolation = Number(getBits(data, 61n, 61n));
  const siloedBorrowingEnabled = Number(getBits(data, 62n, 62n));
  const flashloaningEnabled = Number(getBits(data, 63n, 63n));
  const reserveFactor = getBits(data, 64n, 79n);
  const borrowCap = getBits(data, 80n, 115n);
  const supplyCap = getBits(data, 116n, 151n);
  const liquidationProtocolFee = getBits(data, 152n, 167n);
  const eModeCategory = getBits(data, 168n, 175n);
  const unbackedMintCap = getBits(data, 176n, 211n);
  const debtCeiling = getBits(data, 212n, 251n);

  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    stableRateBorrowingEnabled,
    paused,
    borrowingInIsolation,
    reserveFactor,
    borrowCap,
    supplyCap,
    liquidationProtocolFee,
    eModeCategory,
    unbackedMintCap,
    debtCeiling,
    siloedBorrowingEnabled: !!siloedBorrowingEnabled,
    flashloaningEnabled: !!flashloaningEnabled,
  };
}
