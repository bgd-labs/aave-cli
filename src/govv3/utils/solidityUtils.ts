import { getAddress } from 'viem';
import { TenderlySimulationResponse } from '../../utils/tenderlyClient';

// --- Helper methods ---
/**
 * @notice Given a Tenderly contract object, generates a descriptive human-friendly name for that contract
 * @param contract Tenderly contract object to generate name from
 */
export function getContractName(contracts: TenderlySimulationResponse['contracts'] = [], address: string): string {
  // lower-case comparison as strict equality would fail on child contracts which are lower-cased when returned from tenderly
  const contract = contracts.find((c) => c.address.toLowerCase() === address.toLowerCase());
  if (!contract) return `unknown contract name at \`${getAddress(address)}\``;
  let contractName = contract?.contract_name;

  // If the contract is a token, include the full token name. This is useful in cases where the
  // token is a proxy, so the contract name doesn't give much useful information
  if (contract?.token_data?.name) contractName += ` (${contract?.token_data?.name})`;

  // if the contract is a proxy include it's child
  if (contract.standards?.includes('eip1967') && contract.child_contracts?.[0].address) {
    return `${contractName} at \`${getAddress(contract.address)}\` with implementation ${getContractName(
      contracts,
      contract.child_contracts?.[0].address
    )}`;
  }

  // Lastly, append the contract address and save it off
  return `${contractName} at \`${getAddress(contract.address)}\``;
}
