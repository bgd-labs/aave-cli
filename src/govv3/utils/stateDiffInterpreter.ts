import { Hex, PublicClient, getContract } from 'viem';
import { tenderlyDeepDiff } from './tenderlyDeepDiff';
import { ERC20_ABI } from '../abis/ERC20';
import * as pools from '@bgd-labs/aave-address-book';
import { getBits } from '../../utils/storageSlots';

export async function interpretStateChange(
  contractAddress: string,
  name: string = '',
  original: Record<string, any>,
  dirty: Record<string, any>,
  key: Hex,
  publicClient: PublicClient
) {
  if (name === '_reserves' && (original.configuration.data || dirty.configuration.data))
    return await reserveConfigurationChanged(contractAddress, original, dirty, key, publicClient);
  return undefined;
}

async function reserveConfigurationChanged(
  contractAddress: string,
  original: Record<string, any>,
  dirty: Record<string, any>,
  key: Hex,
  publicClient: PublicClient
) {
  const configurationBefore = getDecodedReserveData(contractAddress, original.configuration.data);
  const configurationAfter = getDecodedReserveData(contractAddress, dirty.configuration.data);
  let symbol = 'unknown';
  try {
    const erc20Contract = getContract({ publicClient, address: key, abi: ERC20_ABI });
    symbol = await erc20Contract.read.symbol();
  } catch (e) {}
  // const symbol =
  return `# decoded configuration.data for key \`${key}\` (symbol: ${symbol})
  ${tenderlyDeepDiff(configurationBefore, configurationAfter, 'configuration.data')}`;
}

function getDecodedReserveData(contractAddress: string, data?: bigint) {
  if (!data) return {};
  if (
    [pools.AaveV2EthereumAMM.POOL, pools.AaveV2Ethereum.POOL, pools.AaveV2Polygon.POOL, pools.AaveV2Avalanche.POOL]
      .map((address) => address.toLowerCase())
      .includes(contractAddress.toLowerCase())
  )
    return decodeReserveDataV2(data);
  return decodeReserveDataV3(data);
}

function decodeReserveDataV2(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const reserveFactor = getBits(data, 64n, 79n);
  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    reserveFactor,
  };
}

function decodeReserveDataV3(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const stableRateBorrowingEnabled = Number(getBits(data, 59n, 59n));
  const paused = Number(getBits(data, 60n, 60n));
  const borrowingInIsolation = Number(getBits(data, 61n, 61n));
  const siloedBorrowingEnabled = Number(getBits(data, 62n, 62n));
  const flashloaningEnabled = Number(getBits(data, 63n, 63n));
  const reserveFactor = getBits(data, 64n, 79n);
  const borrowCap = getBits(data, 80n, 115n);
  const supplyCap = getBits(data, 116n, 151n);
  const liquidationProtocolFee = getBits(data, 152n, 167n);
  const eModeCategory = getBits(data, 168n, 175n);
  const unbackedMintCap = getBits(data, 176n, 211n);
  const debtCeiling = getBits(data, 212n, 251n);

  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    stableRateBorrowingEnabled,
    paused,
    borrowingInIsolation,
    reserveFactor,
    borrowCap,
    supplyCap,
    liquidationProtocolFee,
    eModeCategory,
    unbackedMintCap,
    debtCeiling,
    siloedBorrowingEnabled: !!siloedBorrowingEnabled,
    flashloaningEnabled: !!flashloaningEnabled,
  };
}
