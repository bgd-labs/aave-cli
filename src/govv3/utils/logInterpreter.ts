import {type Client, type Hex, formatUnits} from 'viem';
import type {Input} from '../../utils/tenderlyClient';
import {findAsset} from './checkAddress';
import {formatNumberString, prettifyNumber} from './markdownUtils';

// events emitted typically on the erc20
const tokenAmountEvents = ['Transfer', 'Approval', 'Burn', 'Mint', 'BalanceTransfer'];

// events emitted on the pool
const reserveEvents = ['Withdraw', 'Supply', 'Deposit'];

export async function interpretLog(
  client: Client,
  address: Hex,
  name: string | null,
  inputs: Input[],
) {
  if (name && tokenAmountEvents.includes(name)) {
    // fields formatted by the asset decimal
    const decimalFieldNames = ['value', 'amount', 'wad'];
    for (const name of decimalFieldNames) {
      const valueIndex = inputs.findIndex((i) => i.soltype!.name === name);
      if (valueIndex !== -1) {
        const asset = await findAsset(client, address);
        if (asset) {
          inputs[valueIndex].value = prettifyNumber({
            value: inputs[valueIndex].value as string,
            decimals: asset.decimals,
          });
        }
      }
    }
  }
  if (name && reserveEvents.includes(name)) {
    const valueIndex = inputs.findIndex((i) => i.soltype!.name === 'amount');
    const reserveIndex = inputs.findIndex((i) => i.soltype!.name === 'reserve');
    if (valueIndex !== -1 && reserveIndex !== -1) {
      const asset = await findAsset(client, inputs[reserveIndex].value as Hex);
      if (asset) {
        inputs[valueIndex].value = prettifyNumber({
          value: inputs[valueIndex].value as string,
          decimals: asset.decimals,
        });
      }
    }
  }
  if (name && ['Mint', 'ReserveDataUpdated', 'Burn'].includes(name)) {
    // fields formatted by the asset decimal
    const decimalFieldNames = ['liquidityIndex', 'variableBorrowIndex', 'index'];
    for (const name of decimalFieldNames) {
      const valueIndex = inputs.findIndex((i) => i.soltype!.name === name);
      if (valueIndex !== -1) {
        inputs[valueIndex].value = prettifyNumber({
          value: inputs[valueIndex].value as string,
          decimals: 27,
        });
      }
    }
  }
  const parsedInputs = inputs
    ?.map(
      (i) =>
        `${i.soltype!.name}: ${typeof i.value === 'object' ? JSON.stringify(i.value) : i.value}`,
    )
    .join(', ');
  return `  - \`${name}(${parsedInputs || ''})\``;
}
