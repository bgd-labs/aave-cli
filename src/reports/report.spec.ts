import { describe, it, expect } from 'vitest';
import { diffReports } from './diff-reports';
import { readJsonFile } from '../utils/json';

describe('report', () => {
  it('should generate a well formatted report', async () => {
    const from = readJsonFile('/src/mocks/preTestEngineArbV3.json');
    const to = readJsonFile('/src/mocks/postTestEngineArbV3.json');
    const content = await diffReports(from, to);
    console.log(content);
    expect(content).toMatchSnapshot();
  });
});
