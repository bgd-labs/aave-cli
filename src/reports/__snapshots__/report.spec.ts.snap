// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`report > should generate a well formatted report 1`] = `
"## Reserve changes

### Reserves added

#### 1INCH ([0x9c2C5fd7b07E95EE044DDeba0E97a665F142394f](https://polygonscan.com/address/0x9c2C5fd7b07E95EE044DDeba0E97a665F142394f))

| description | value |
| --- | --- |
| decimals | 18 |
| isActive | true |
| isFrozen | false |
| supplyCap | 85,000 1INCH |
| borrowCap | 60,000 1INCH |
| debtCeiling | 0 $ |
| isSiloed | false |
| isFlashloanable | false |
| eModeCategory | 0 |
| oracle | [0x443C5116CdF663Eb387e72C688D276e702135C87](https://polygonscan.com/address/0x443C5116CdF663Eb387e72C688D276e702135C87) |
| oracleDecimals | 8 |
| oracleDescription | 1INCH / USD |
| oracleLatestAnswer | 0.52690564 |
| usageAsCollateralEnabled | true |
| ltv | 82.5 % |
| liquidationThreshold | 86 % |
| liquidationBonus | 5 % |
| liquidationProtocolFee | 10 % |
| reserveFactor | 10 % |
| aToken | [0xf59036CAEBeA7dC4b86638DFA2E3C97dA9FcCd40](https://polygonscan.com/address/0xf59036CAEBeA7dC4b86638DFA2E3C97dA9FcCd40) |
| aTokenImpl | [0xa5ba6E5EC19a1Bf23C857991c857dB62b2Aa187B](https://polygonscan.com/address/0xa5ba6E5EC19a1Bf23C857991c857dB62b2Aa187B) |
| variableDebtToken | [0x77fA66882a8854d883101Fb8501BD3CaD347Fc32](https://polygonscan.com/address/0x77fA66882a8854d883101Fb8501BD3CaD347Fc32) |
| variableDebtTokenImpl | [0x81387c40EB75acB02757C1Ae55D5936E78c9dEd3](https://polygonscan.com/address/0x81387c40EB75acB02757C1Ae55D5936E78c9dEd3) |
| stableDebtToken | [0x173e54325AE58B072985DbF232436961981EA000](https://polygonscan.com/address/0x173e54325AE58B072985DbF232436961981EA000) |
| stableDebtTokenImpl | [0x52A1CeB68Ee6b7B5D13E0376A1E0E4423A8cE26e](https://polygonscan.com/address/0x52A1CeB68Ee6b7B5D13E0376A1E0E4423A8cE26e) |
| borrowingEnabled | true |
| stableBorrowRateEnabled | false |
| isBorrowableInIsolation | false |
| interestRateStrategy | [0x03733F4E008d36f2e37F0080fF1c8DF756622E6F](https://polygonscan.com/address/0x03733F4E008d36f2e37F0080fF1c8DF756622E6F) |
| optimalUsageRatio | 45 % |
| maxExcessUsageRatio | 55 % |
| baseVariableBorrowRate | 0 % |
| variableRateSlope1 | 7 % |
| variableRateSlope2 | 300 % |
| baseStableBorrowRate | 9 % |
| stableRateSlope1 | 0 % |
| stableRateSlope2 | 0 % |
| optimalStableToTotalDebtRatio | 20 % |
| maxExcessStableToTotalDebtRatio | 80 % |
| interestRate | ![ir](/.assets/4ab0dbcff3f330539d66319942d38435e45137bf.svg) |

## Raw diff

\`\`\`json
{
  \\"reserves\\": {
    \\"0x9c2C5fd7b07E95EE044DDeba0E97a665F142394f\\": {
      \\"from\\": null,
      \\"to\\": {
        \\"aToken\\": \\"0xf59036CAEBeA7dC4b86638DFA2E3C97dA9FcCd40\\",
        \\"aTokenImpl\\": \\"0xa5ba6E5EC19a1Bf23C857991c857dB62b2Aa187B\\",
        \\"borrowCap\\": 60000,
        \\"borrowingEnabled\\": true,
        \\"debtCeiling\\": 0,
        \\"decimals\\": 18,
        \\"eModeCategory\\": 0,
        \\"interestRateStrategy\\": \\"0x03733F4E008d36f2e37F0080fF1c8DF756622E6F\\",
        \\"isActive\\": true,
        \\"isBorrowableInIsolation\\": false,
        \\"isFlashloanable\\": false,
        \\"isFrozen\\": false,
        \\"isSiloed\\": false,
        \\"liquidationBonus\\": 10500,
        \\"liquidationProtocolFee\\": 1000,
        \\"liquidationThreshold\\": 8600,
        \\"ltv\\": 8250,
        \\"oracle\\": \\"0x443C5116CdF663Eb387e72C688D276e702135C87\\",
        \\"oracleDecimals\\": 8,
        \\"oracleDescription\\": \\"1INCH / USD\\",
        \\"oracleLatestAnswer\\": 52690564,
        \\"reserveFactor\\": 1000,
        \\"stableBorrowRateEnabled\\": false,
        \\"stableDebtToken\\": \\"0x173e54325AE58B072985DbF232436961981EA000\\",
        \\"stableDebtTokenImpl\\": \\"0x52A1CeB68Ee6b7B5D13E0376A1E0E4423A8cE26e\\",
        \\"supplyCap\\": 85000,
        \\"symbol\\": \\"1INCH\\",
        \\"underlying\\": \\"0x9c2C5fd7b07E95EE044DDeba0E97a665F142394f\\",
        \\"usageAsCollateralEnabled\\": true,
        \\"variableDebtToken\\": \\"0x77fA66882a8854d883101Fb8501BD3CaD347Fc32\\",
        \\"variableDebtTokenImpl\\": \\"0x81387c40EB75acB02757C1Ae55D5936E78c9dEd3\\"
      }
    }
  }
}
\`\`\`"
`;
